src/pages/AddMembersPage/NewMemberList.tsx

```diff
@@ -0,0 +1,269 @@
+import { normalizeError } from '../../utils/reporting';
+
+import CopyCodeTextField from './CopyCodeTextField';
+import NewMemberEntry from './NewMemberEntry';
+import NewMemberGridBox from './NewMemberGridBox';
+
+const NewMemberList = ({
+  // TODO: figure out what to do w/ revoke
+  // revokeWelcome,
+  circleId,
+  welcomeLink,
+}: {
+  circleId: number;
+  welcomeLink: string;
+  revokeWelcome(): void;
+}) => {
+  const { fetchCircle } = useApiBase();
+  const { showError } = useApeSnackbar();
+
+  const [loading, setLoading] = useState<boolean>();
+  const [successCount, setSuccessCount] = useState<number>(0);
+
+  const successRef = useRef<HTMLDivElement>(null);
+
+  const queryClient = useQueryClient();
+
+  const newMemberSchema = z.object({
+    newMembers: z
+      .array(
+        z
+          .object({
+            address: zEthAddress.or(z.literal('')),
+            name: z
+              .string()
+              .min(3, 'Name must be at least 3 characters')
+              .or(z.literal('')),
+          })
+          .superRefine((data, ctx) => {
+            if (data.name && data.name !== '' && data.address === '') {
+              ctx.addIssue({
+                code: z.ZodIssueCode.custom,
+                path: ['address'],
+                message: 'Address is required if name is provided',
+              });
+            }
+            if (data.address && data.address !== '' && data.name === '') {
+              ctx.addIssue({
+                code: z.ZodIssueCode.custom,
+                path: ['name'],
+                message: 'Name is required if address is provided',
+              });
+            }
+          })
+      )
+      .superRefine(data => {
+        if (data.filter(m => m.address != '' && m.name != '').length == 0) {
+          //TODO: I want to use this to prevent form submission of there are no valid entries
+          // but it just breaks the more useful errors from rendering
+          // ctx.addIssue({
+          //   code: z.ZodIssueCode.custom,
+          //   message: 'no valid members entered',
+          // });
+        }
+      }),
+  });
+
+  const defaultValues = {
+    newMembers: [
+      { name: '', address: '' },
+      { name: '', address: '' },
+    ],
+  };
+
+  const { register, control, handleSubmit, reset, formState, getValues } =
+    useForm({
+      resolver: zodResolver(newMemberSchema),
+      reValidateMode: 'onChange',
+      mode: 'onChange',
+      defaultValues,
+    });
+  const { errors } = formState;
+
+  const {
+    fields: newMemberFields,
+    append: appendNewMember,
+    remove: removeNewMember,
+  } = useFieldArray({
+    name: 'newMembers',
+    control,
+  });
+
+  const submitNewMembers = async () => {
+    const newMembers = getValues('newMembers') as {
+      name: string;
+      address: string;
+    }[];
+    try {
+      setLoading(true);
+      setSuccessCount(0);
+      const filteredMembers = newMembers.filter(
```

Comment:
Will you need to filter for duplicate name or addresses here ?

Reply:
I dont think we do , this happens on the API side and it returns a decent error message

Reply:
I improved it a little bit